Added a consensus RFC (#997)
tari-project/tari: RFC/src/RFC-0120_Consensus.md

technological merits of the potential system outlined herein.

## Goals

The aim of this Request for Comment (RFC) is to describe the fields that a block should contain as well as all consensus rules that will determine the validity of a block.

## Related Requests for Comment

* [RFC-0100: Base Layer](RFC-0100_BaseLayer.md)
* [RFC-0130: Mining](./RFC-0130_Mining.md)
* [RFC-0140: SyncAndSeeding](RFC-0140_Syncing_and_seeding.md)

## Description

### Blocks

Every [block] MUST conform to the following:

* Have one and only valid coinbase [UTXO] and kernel
* have one and only valid blockheader;
* [cut-through] has been applied where possible;
* every [UTXO] has a valid [range proof].

If a [block] does not confirm to the above, the block should be rejected as invalid and the peer from which it was received marked as a malicious peer.

### Blockheaders

Every [block header] MUST contain the following fields:

* version;
* height;
* prev_hash;
* timestamp;
* output_mr;
* range_proof_mr;
* kernel_mr;
* total_kernel_offset;
* total_difficulty;
* nonce;
* pow.

The [block header] MUST confirm to the following:

> * The nonce and pow must be a valid for the [block header].

Author: There is no PoW in the glossary but will add that

* All the merkle roots must be valid for the states after the [block] was applied to the local state.

If the [block header] does not confirm to any of the above, the [block] needs to be rejected and the peer the it was received from, marked as a malicious peer.

#### Version

This is the version currently running on the chain.

The version MUST confirm to the following:

* Must be transmitted as a u16.

> * Current version is 1.

Author: My vote will be for 1

#### Height

This is the counter how much blocks has passed since the genesis block.

The height MUST confirm to the following:

* Must be transmitted as a u64.
* The height must increase be exactly 1 above the previous block.
* Must start at 1 which is the genesis block.

#### Prev_hash

This is the hash of the previous block's header.

The prev_hash MUST confirm to the following:

* Must be transmitted as an array of u8.

> * Must be the hash of the previous block's header.

Author: We should, will add

* Must include the entire previous block's header.

#### Timestamp

This is the timestamp at which the block was mined.

The timestamp MUST confirm to the following:

> * Must be transmitted as <TO BE CONFIRMED>.

Author: So we are sending the timestamp as unix timestamp. I would vote for removal of the nano precision thou. 
The reason I left it out is that I think currently we are sending the timestamp as a chrono datetime<UTC> serialised

> * Must be transmitted as <TO BE CONFIRMED>.

Reviewer: Agree that the nano precision is unnecessary (esp in this case) and could be set to 0 when converted to a wire format. The Timestamp proto file is a standard from Google, so we've made use of it in the proto files. For RFC purposes I think saying a UNIX epoch timestamp will be correct. 

We are serializing blocks in the chain db as a chrono datetime, but over the wire it's a `Timestamp`

> * Must be transmitted as <TO BE CONFIRMED>.

Reviewer: Suppose it really doesn't matter which format an implementer stores the timestamp as long as their wire format is correct. Once the comms RFCs are updated these will provide reference to the wire format types. (basically "just use these proto files and you'll be able to represent a block over the wire"). For this RFC I would include just enough info for the reader to know how to represent the field in whatever language they're coding in.

* Must be less than [FTL].
* Must be higher than the [MTP]

#### Output_mr

This is the merkle root of the outputs. This is calculated in the following way: Hash (txo MMR root  || roaring bitmap hash of UTXO indices).

The output_mr MUST confirm to the following:

* Must be transmitted as an array of [u8].
* The hashing function used must be blake2b with a 256 bit digest.

#### Range_proof_mr

This is the merkle root of the range proofs.

The range_proof_mr MUST confirm to the following:

> * Must be transmitted as an array of [u8].

Author: RFC 170 is a good place to put that I think, and discuss it

> * Must be transmitted as an array of [u8].

Reviewer: Agreed

* The hashing function used must be blake2b with a 256 bit digest.

#### Kernel_mr

This is the merkle root of the outputs.

The kernel_mr MUST confirm to the following:.

* Must be transmitted as an array of [u8].
* The hashing function used must be blake2b with a 256 bit digest.

#### Total_kernel_offset

This is total summed offset of all the transactions contain in this block.

The total_kernel_offset MUST confirm to the following:

* Must be transmitted as <TO BE CONFIRMED>

#### Total_difficulty

This is the total accumulated difficulty of the mined chained.

The total_difficulty MUST confirm to the following:

* Must be transmitted as <TO BE CONFIRMED>;
* Must increase by at least 1 above the previous block;

#### Nonce

This is the nonce used in solving the Proof of Work. 

The nonce MUST confirm to the following:

> * Must be transmitted as <TO BE CONFIRMED>;

Author: This might change. But will put u64 in now, as that for test-net. Main-net will most likely change depending on the PoW we settle on.

> * Must be transmitted as <TO BE CONFIRMED>;

Reviewer: Ah ok - maybe leave it as TBC until it's finalized

> * Must be transmitted as <TO BE CONFIRMED>;

Reviewer 2: The nonce will definitely be a u64.
If we need more degrees of freedom, we can introduce a nonce_2, but we won't widen the nonce field

[BCP 14](https://tools.ietf.org/html/bcp14) (covering RFC2119 and RFC8174) when, and only when, they appear in all capitals, as 
shown here.

## Disclaimer

This document and its content are intended for information purposes only and may be subject to change or update
without notice.

This document may include preliminary concepts that may or may not be in the process of being developed by the Tari
community. The release of this document is intended solely for review and discussion by the community of the
technological merits of the potential system outlined herein.

## Goals

The aim of this Request for Comment (RFC) is to describe the fields that a block should contain as well as all consensus rules that will determine the validity of a block.

## Related Requests for Comment

* [RFC-0100: Base Layer](RFC-0100_BaseLayer.md)
* [RFC-0130: Mining](./RFC-0130_Mining.md)
* [RFC-0140: SyncAndSeeding](RFC-0140_Syncing_and_seeding.md)

## Description

### Blocks

Every [block] MUST conform to the following:

* Have a _single_ valid coinbase [UTXO] and kernel
* have a _single_ valid blockheader;

> * [cut-through] has been applied where possible;

Reviewer 2: ```suggestion
* [cut-through] SHOULD have been applied;
```

Author: Other RFC mentions MUST

Author: MUST also increases privacy for the baselayer

* every [UTXO] has a valid [range proof].

If a [block] does not confirm to the above, the block should be rejected as invalid and the peer from which it was received marked as a malicious peer.

### Blockheaders

Every [block header] MUST contain the following fields:

* version;
* height;
* prev_hash;
* timestamp;
* output_mr;
* range_proof_mr;
* kernel_mr;
* total_kernel_offset;
* total_difficulty;
* nonce;
* pow.

The [block header] MUST confirm to the following:

> * The nonce and [PoW](#pow) must be a valid for the [block header].

Reviewer 2: ```suggestion
* The nonce and [PoW](#pow) must be valid for the [block header].
```

> * All the merkle roots must be valid for the states after the [block] was applied to the local state.

Reviewer 2: ```suggestion
* All the merkle roots must be valid for the states _after_ the [block] was applied to the local state.
```

> If the [block header] does not confirm to any of the above, the [block] needs to be rejected and the peer the it was received from, marked as a malicious peer.

Reviewer 2: ```suggestion
If the [block header] does not confirm to any of the above, the [block] MUST be rejected.
If a peer rejects multiple blocks from a given peer, it MAY blacklist the peer and ignore further communication from it.
```

#### Version

This is the version currently running on the chain.

The version MUST confirm to the following:

* It is represented as an unsigned 16-bit integer.

> * Version numbers should be incremented whenever there is a change in the blockchain schema starting from 1.

Reviewer 2: ```suggestion
* Version numbers MUST be incremented whenever there is a change in the blockchain schema starting from 1.
```

#### Height

> A counter indicating how much blocks have passed since the genesis block (inclusive).

Reviewer 2: ```suggestion
A counter indicating how many blocks have passed since the genesis block (inclusive).
```

The height MUST confirm to the following:

* Represented as an unsigned 64-bit integer.
* The height MUST be exactly 1 more than the block referenced in the `prev_hash` block header field.

> * The genesis block MUST have a height of 1.

Reviewer 2: ```suggestion
* The genesis block MUST have a height of 0.
```

We actually have Genblock at 0. 

I can't think of any reason to change this.

This is the timestamp at which the block was mined.

The timestamp MUST confirm to the following:

* Must be transmitted as UNIX timestamp.
* MUST be less than [FTL].
* MUST be higher than the [MTP].

#### Output_mr

This is the merkle root of the outputs. This is calculated in the following way: Hash (txo MMR root  || roaring bitmap hash of UTXO indices).

The output_mr MUST confirm to the following:

* Represented as an array of unsigned 8-bit integers (bytes).
* The hashing function used MUST be blake2b with a 256 bit digest.

#### Range_proof_mr

This is the merkle root of the range proofs.

The range_proof_mr MUST confirm to the following:

* Represented as an array of unsigned 8-bit integers (bytes).
* The hashing function used must be blake2b with a 256 bit digest.

#### Kernel_mr

This is the merkle root of the outputs.

The kernel_mr MUST confirm to the following:.

* Must be transmitted as an array of unsigned 8-bit integers (bytes).
* The hashing function used must be blake2b with a 256 bit digest.

#### Total_kernel_offset

This is total summed offset of all the transactions contained in this block.

The total_kernel_offset MUST confirm to the following:

> * Must be transmitted as an array of unsigned 8-bit integers (bytes)

Reviewer 2: BTW, all of these fields are in little-endian format.

#### Total_difficulty

This is the total accumulated difficulty of the mined chained.

The total_difficulty MUST confirm to the following:

* Must be transmitted as unsigned 64-bit integer;

> * MUST be larger than the previous block's `total_difficulty`;

Reviewer 2: * meet the difficulty target for the block as determined by the consensus difficulty algorithm

#### Nonce

This is the nonce used in solving the Proof of Work. 

The nonce MUST confirm to the following:

* Must be transmitted as unsigned 64-bit integer;

#### PoW

This is Proof of Work algorithm that was used to solve the Proof of Work. This is used in conjunction with the Nonce

The [PoW] MUST confirm to the following:

* Must be transmitted as <TO BE CONFIRMED>;

### FTL

The Future Time Limit. This is how far into the future a time is accepted as a valid time. Any time that is more than the FTL is rejected until such a time that it is not less than the FTL.
The FTL is calculated as (T*N)/20 with T and N defined as:
T: Target time - This is the ideal time that should pass between blocks which have been mined.
N: Block window - This is the amount of blocks used when calculating difficulty adjustments.

### MTP

The Median Time Past. This is the lower limit of a time. Any time that is less than the MTP is rejected.

> THe MTP is calculated as the average timestamp of the previous 11 blocks.

Reviewer 2: Median is surely better.
Average is not robust and could be gamed by malicious miners.
